File: frontend/mobile/lib/core/services/qibla_service.dart
import 'dart:async';
import 'dart:math' as math;
import 'package:flutter_compass/flutter_compass.dart';
import 'package:geolocator/geolocator.dart';
import 'package:sensors_plus/sensors_plus.dart';
import '../../../features/qibla/data/models/qibla_model.dart';

/// Service for calculating Qibla direction and managing compass functionality
class QiblaService {
  // Kaaba coordinates (Mecca, Saudi Arabia)
  static const double kaabaLatitude = 21.4225;
  static const double kaabaLongitude = 39.8262;

  StreamSubscription<CompassEvent>? _compassSubscription;
  StreamSubscription<MagnetometerEvent>? _magnetometerSubscription;

  /// Calculate Qibla direction from current location
  Future<QiblaModel> calculateQiblaDirection(Position position) async {
    final qiblaDirection = _calculateBearing(
      position.latitude,
      position.longitude,
      kaabaLatitude,
      kaabaLongitude,
    );

    final distance = _calculateDistance(
      position.latitude,
      position.longitude,
      kaabaLatitude,
      kaabaLongitude,
    );

    // Get location name (simplified - in production, use geocoding service)
    final locationName = 'Lat: ${position.latitude.toStringAsFixed(4)}, '
        'Lon: ${position.longitude.toStringAsFixed(4)}';

    return QiblaModel(
      direction: qiblaDirection,
      distance: distance,
      locationName: locationName,
      latitude: position.latitude,
      longitude: position.longitude,
      calculatedAt: DateTime.now(),
    );
  }

  /// Get current device location
  Future<Position> getCurrentLocation() async {
    bool serviceEnabled;
    LocationPermission permission;

    // Check if location services are enabled
    serviceEnabled = await Geolocator.isLocationServiceEnabled();
    if (!serviceEnabled) {
      throw Exception('خدمات الموقع غير مفعلة. يرجى تفعيل GPS.');
    }

    // Check location permissions
    permission = await Geolocator.checkPermission();
    if (permission == LocationPermission.denied) {
      permission = await Geolocator.requestPermission();
      if (permission == LocationPermission.denied) {
        throw Exception('تم رفض إذن الوصول إلى الموقع.');
      }
    }

    if (permission == LocationPermission.deniedForever) {
      throw Exception(
        'إذن الوصول إلى الموقع مرفوض بشكل دائم. '
        'يرجى تفعيله من إعدادات التطبيق.',
      );
    }

    // Get current position
    return await Geolocator.getCurrentPosition(
      desiredAccuracy: LocationAccuracy.high,
    );
  }

  /// Stream of compass heading updates
  Stream<double> getCompassHeading() {
    return FlutterCompass.events!.map((event) {
      // Normalize heading to 0-360 range
      double heading = event.heading ?? 0.0;
      if (heading < 0) heading += 360;
      return heading;
    });
  }

  /// Check if compass is available on device
  Future<bool> isCompassAvailable() async {
    final compassEvents = FlutterCompass.events;
    return compassEvents != null;
  }

  /// Calculate compass calibration status based on magnetometer data
  Stream<CompassCalibration> getCalibrationStatus() async* {
    await for (final event in magnetometerEvents) {
      // Calculate magnetic field strength
      final x = event.x;
      final y = event.y;
      final z = event.z;
      final magnitude = math.sqrt(x * x + y * y + z * z);

      // Typical Earth's magnetic field is 25-65 microteslas
      // If magnitude is too low or too high, compass needs calibration
      if (magnitude < 20 || magnitude > 70) {
        yield CompassCalibration(
          isCalibrated: false,
          accuracy: _calculateAccuracy(magnitude),
          message: 'يرجى معايرة البوصلة بتحريك الجهاز على شكل رقم 8',
        );
      } else {
        yield CompassCalibration(
          isCalibrated: true,
          accuracy: _calculateAccuracy(magnitude),
          message: 'البوصلة معايرة بشكل صحيح',
        );
      }
    }
  }

  /// Calculate bearing (direction) from one point to another
  /// Returns angle in degrees (0-360) where 0 is North
  double _calculateBearing(
    double lat1,
    double lon1,
    double lat2,
    double lon2,
  ) {
    final dLon = _toRadians(lon2 - lon1);
    final lat1Rad = _toRadians(lat1);
    final lat2Rad = _toRadians(lat2);

    final y = math.sin(dLon) * math.cos(lat2Rad);
    final x = math.cos(lat1Rad) * math.sin(lat2Rad) -
        math.sin(lat1Rad) * math.cos(lat2Rad) * math.cos(dLon);

    final bearing = math.atan2(y, x);
    final bearingDegrees = _toDegrees(bearing);

    // Normalize to 0-360
    return (bearingDegrees + 360) % 360;
  }

  /// Calculate distance between two points using Haversine formula
  /// Returns distance in kilometers
  double _calculateDistance(
    double lat1,
    double lon1,
    double lat2,
    double lon2,
  ) {
    const earthRadius = 6371.0; // Earth's radius in kilometers

    final dLat = _toRadians(lat2 - lat1);
    final dLon = _toRadians(lon2 - lon1);

    final a = math.sin(dLat / 2) * math.sin(dLat / 2) +
        math.cos(_toRadians(lat1)) *
            math.cos(_toRadians(lat2)) *
            math.sin(dLon / 2) *
            math.sin(dLon / 2);

    final c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a));

    return earthRadius * c;
  }

  /// Calculate accuracy score based on magnetic field magnitude
  double _calculateAccuracy(double magnitude) {
    // Ideal range is 25-65 microteslas
    const idealMin = 25.0;
    const idealMax = 65.0;
    const idealMid = (idealMin + idealMax) / 2;

    if (magnitude >= idealMin && magnitude <= idealMax) {
      // Within ideal range - calculate how close to center
      final deviation = (magnitude - idealMid).abs();
      final maxDeviation = (idealMax - idealMin) / 2;
      return 1.0 - (deviation / maxDeviation) * 0.3; // 0.7 to 1.0
    } else if (magnitude < idealMin) {
      // Below ideal range
      return (magnitude / idealMin).clamp(0.0, 0.7);
    } else {
      // Above ideal range
      final excess = magnitude - idealMax;
      return (1.0 - (excess / idealMax)).clamp(0.0, 0.7);
    }
  }

  double _toRadians(double degrees) => degrees * math.pi / 180.0;
  double _toDegrees(double radians) => radians * 180.0 / math.pi;

  /// Clean up resources
  void dispose() {
    _compassSubscription?.cancel();
    _magnetometerSubscription?.cancel();
  }
}

File: frontend/nextjs-app/src/lib/services/qibla-service.ts
import { QiblaData, GeolocationPosition, CompassCalibration } from '@/types/qibla';

/**
 * Service for calculating Qibla direction and managing compass functionality
 */
export class QiblaService {
  // Kaaba coordinates (Mecca, Saudi Arabia)
  private static readonly KAABA_LATITUDE = 21.4225;
  private static readonly KAABA_LONGITUDE = 39.8262;

  /**
   * Calculate Qibla direction from current location
   */
  static calculateQiblaDirection(position: GeolocationPosition): QiblaData {
    const qiblaDirection = this.calculateBearing(
      position.latitude,
      position.longitude,
      this.KAABA_LATITUDE,
      this.KAABA_LONGITUDE
    );

    const distance = this.calculateDistance(
      position.latitude,
      position.longitude,
      this.KAABA_LATITUDE,
      this.KAABA_LONGITUDE
    );

    const locationName = `Lat: ${position.latitude.toFixed(4)}, Lon: ${position.longitude.toFixed(4)}`;

    return {
      direction: qiblaDirection,
      distance,
      locationName,
      latitude: position.latitude,
      longitude: position.longitude,
      calculatedAt: new Date().toISOString(),
    };
  }

  /**
   * Get current device location using Geolocation API
   */
  static async getCurrentLocation(): Promise<GeolocationPosition> {
    return new Promise((resolve, reject) => {
      if (!navigator.geolocation) {
        reject(new Error('Geolocation is not supported by your browser'));
        return;
      }

      navigator.geolocation.getCurrentPosition(
        (position) => {
          resolve({
            latitude: position.coords.latitude,
            longitude: position.coords.longitude,
            accuracy: position.coords.accuracy,
          });
        },
        (error) => {
          let errorMessage = 'Unable to retrieve your location';
          switch (error.code) {
            case error.PERMISSION_DENIED:
              errorMessage = 'Location permission denied';
              break;
            case error.POSITION_UNAVAILABLE:
              errorMessage = 'Location information unavailable';
              break;
            case error.TIMEOUT:
              errorMessage = 'Location request timed out';
              break;
          }
          reject(new Error(errorMessage));
        },
        {
          enableHighAccuracy: true,
          timeout: 10000,
          maximumAge: 0,
        }
      );
    });
  }

  /**
   * Check if device orientation API is available
   */
  static isOrientationAvailable(): boolean {
    return 'DeviceOrientationEvent' in window;
  }

  /**
   * Request permission for device orientation (iOS 13+)
   */
  static async requestOrientationPermission(): Promise<boolean> {
    if (typeof (DeviceOrientationEvent as any).requestPermission === 'function') {
      try {
        const permission = await (DeviceOrientationEvent as any).requestPermission();
        return permission === 'granted';
      } catch (error) {
        console.error('Error requesting orientation permission:', error);
        return false;
      }
    }
    // Permission not needed on non-iOS devices
    return true;
  }

  /**
   * Calculate bearing (direction) from one point to another
   * Returns angle in degrees (0-360) where 0 is North
   */
  private static calculateBearing(
    lat1: number,
    lon1: number,
    lat2: number,
    lon2: number
  ): number {
    const dLon = this.toRadians(lon2 - lon1);
    const lat1Rad = this.toRadians(lat1);
    const lat2Rad = this.toRadians(lat2);

    const y = Math.sin(dLon) * Math.cos(lat2Rad);
    const x =
      Math.cos(lat1Rad) * Math.sin(lat2Rad) -
      Math.sin(lat1Rad) * Math.cos(lat2Rad) * Math.cos(dLon);

    const bearing = Math.atan2(y, x);
    const bearingDegrees = this.toDegrees(bearing);

    // Normalize to 0-360
    return (bearingDegrees + 360) % 360;
  }

  /**
   * Calculate distance between two points using Haversine formula
   * Returns distance in kilometers
   */
  private static calculateDistance(
    lat1: number,
    lon1: number,
    lat2: number,
    lon2: number
  ): number {
    const earthRadius = 6371.0; // Earth's radius in kilometers

    const dLat = this.toRadians(lat2 - lat1);
    const dLon = this.toRadians(lon2 - lon1);

    const a =
      Math.sin(dLat / 2) * Math.sin(dLat / 2) +
      Math.cos(this.toRadians(lat1)) *
        Math.cos(this.toRadians(lat2)) *
        Math.sin(dLon / 2) *
        Math.sin(dLon / 2);

    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

    return earthRadius * c;
  }

  /**
   * Calculate compass calibration status
   */
  static calculateCalibrationStatus(
    magneticField?: { x: number; y: number; z: number }
  ): CompassCalibration {
    if (!magneticField) {
      return {
        isCalibrated: false,
        accuracy: 0,
        message: 'Compass calibration needed',
      };
    }

    const magnitude = Math.sqrt(
      magneticField.x ** 2 + magneticField.y ** 2 + magneticField.z ** 2
    );

    // Typical Earth's magnetic field is 25-65 microteslas
    if (magnitude < 20 || magnitude > 70) {
      return {
        isCalibrated: false,
        accuracy: this.calculateAccuracy(magnitude),
        message: 'Please calibrate compass by moving device in figure-8 pattern',
      };
    }

    return {
      isCalibrated: true,
      accuracy: this.calculateAccuracy(magnitude),
      message: 'Compass calibrated correctly',
    };
  }

  /**
   * Calculate accuracy score based on magnetic field magnitude
   */
  private static calculateAccuracy(magnitude: number): number {
    const idealMin = 25.0;
    const idealMax = 65.0;
    const idealMid = (idealMin + idealMax) / 2;

    if (magnitude >= idealMin && magnitude <= idealMax) {
      const deviation = Math.abs(magnitude - idealMid);
      const maxDeviation = (idealMax - idealMin) / 2;
      return 1.0 - (deviation / maxDeviation) * 0.3;
    } else if (magnitude < idealMin) {
      return Math.min(magnitude / idealMin, 0.7);
    } else {
      const excess = magnitude - idealMax;
      return Math.max(1.0 - excess / idealMax, 0);
    }
  }

  /**
   * Normalize heading to 0-360 range
   */
  static normalizeHeading(heading: number): number {
    let normalized = heading % 360;
    if (normalized < 0) normalized += 360;
    return normalized;
  }

  /**
   * Calculate relative direction from device heading to Qibla
   */
  static calculateRelativeDirection(
    deviceHeading: number,
    qiblaDirection: number
  ): number {
    let diff = qiblaDirection - deviceHeading;

    // Normalize to -180 to 180 range
    if (diff > 180) {
      diff -= 360;
    } else if (diff < -180) {
      diff += 360;
    }

    return diff;
  }

  /**
   * Check if device is pointing towards Qibla
   */
  static isPointingToQibla(relativeDirection: number, tolerance: number = 5): boolean {
    return Math.abs(relativeDirection) <= tolerance;
  }

  /**
   * Get cardinal direction name in Arabic
   */
  static getCardinalDirection(heading: number): string {
    const normalized = this.normalizeHeading(heading);

    if (normalized >= 337.5 || normalized < 22.5) return 'شمال';
    if (normalized >= 22.5 && normalized < 67.5) return 'شمال شرق';
    if (normalized >= 67.5 && normalized < 112.5) return 'شرق';
    if (normalized >= 112.5 && normalized < 157.5) return 'جنوب شرق';
    if (normalized >= 157.5 && normalized < 202.5) return 'جنوب';
    if (normalized >= 202.5 && normalized < 247.5) return 'جنوب غرب';
    if (normalized >= 247.5 && normalized < 292.5) return 'غرب';
    return 'شمال غرب';
  }

  /**
   * Format distance for display
   */
  static formatDistance(distanceKm: number): string {
    if (distanceKm < 1) {
      return `${Math.round(distanceKm * 1000)} متر`;
    } else if (distanceKm < 100) {
      return `${distanceKm.toFixed(1)} كم`;
    } else {
      return `${Math.round(distanceKm)} كم`;
    }
  }

  private static toRadians(degrees: number): number {
    return (degrees * Math.PI) / 180;
  }

  private static toDegrees(radians: number): number {
    return (radians * 180) / Math.PI;
  }
}
